app/demo-loop/KernelUtils.i.hh

```diff
@@ -0,0 +1,107 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2021 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file KernelUtils.i.hh
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include "base/Macros.hh"
+#include "geometry/LinearPropagator.hh"
+#include "random/distributions/ExponentialDistribution.hh"
+
+namespace demo_loop
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Sample mean free path and calculate physics step limits.
+ */
+template<class Rng>
+CELER_FUNCTION void calc_step_limits(const GeoTrackView&      geo,
+                                     const GeoMaterialView&   geo_mat,
+                                     const MaterialTrackView& mat,
+                                     const ParticleTrackView& particle,
+                                     PhysicsTrackView&        phys,
+                                     Rng&                     rng)
+{
+    // Sample mean free path
+    if (!phys.has_interaction_mfp())
+    {
+        ExponentialDistribution<real_type> sample_exponential;
+        phys.interaction_mfp(sample_exponential(rng));
+    }
+
+    // Calculate physics step limits
+    if (particle.is_stopped())
+    {
+        // Set the interaction length to zero for stopped particles
+        phys.step_length(0);
+    }
+    else
+    {
+        real_type step = calc_tabulated_physics_step(mat, particle, phys);
+        phys.step_length(step);
+    }
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Propagate up to the step length or next boundary.
+ */
+CELER_FUNCTION real_type propagate(GeoTrackView&           geo,
+                                   const PhysicsTrackView& phys)
+{
+    // Actual distance, limited by along-step length or geometry
+    real_type step = phys.step_length();
+    if (step > 0)
+    {
+        // Propagate up to the step length or next boundary
+        LinearPropagator propagate(&geo);
+        auto             geo_step = propagate(step);
+        step                      = geo_step.distance;
```

Comment:
Here we'll need to add code to check whether the volume/material have changed.

Reply:
Added a reminder :)